### DISCUSSION TITLE
Questions about placing pins manually

### DISCUSSION THREAD

**Discussion by b224hisl** [The-OpenROAD-Project/OpenROAD-flow-scripts#1690] [Q&A] [answered]
Hi, I'm trying to place pins manually after the floorplan stage, because I don't want to place pins automatically around the die boundary. 
I notice that there are some commands related to that. But I'm quite confused about these commands.

1. What's the difference between `define_pin_shape_pattern ` and `set_io_pin_constraint`
2. There are two commands called` place_pins` and `place_pin`, one for placing all pins and the other one is used to place a specific pin. But the later one just has one option `-layer`, the former one has `-hor_layers` and `-ver_layers`. I'm curious why placing pins needs to define hor/ver layers? In my understanding, when placing a pin we just need to define which metal layer the pin is on.
3. For placing pins manually, what's the right order of executing these commands?

Looking forward to your reply, thank you very very much!

**Comment by rovinski** [comment, marked as answer]
1. `define_pin_shape_pattern` creates a regular grid where pins can be placed throughout an area. This is mostly used if you want to create pins on your top layer, such as for a block which uses M1-M4 so that pins can be connected on M4 from above. The much more common use case is for advanced packaging. From the README: The define_pin_shape_pattern command can be used to place pins in any metal layer with the minimum allowed spacing to facilitate 3DIC integration of chips using face-to-face packaging technologies. These technologies include [micro bumps](https://semiengineering.com/bumps-vs-hybrid-bonding-for-advanced-packaging/) and [hybrid bonding](https://www.3dincites.com/2018/04/hybrid-bonding-from-concept-to-commercialization/) for high density face-to-face interconnect.
`set_io_pin_constraint` assumes the normal pin placement rules, where pins will be created only on the die edges and on the routing tracks. The purpose of the command is to restrict where the pin placer will place your pins, such as to specific metal layers or to specific sides of the die. This command is used more often with blocks.
2. `place_pins` is intended to use the pin placement engine to automatically place pins in valid locations (as determined by the metal layers, routing tracks, and constraints set with `set_io_pin_constraint`).
`place_pin` is a manual command intended to place a pin at an exact location without using the pin placement engine. `place_pin` has a `-layer` option because there is only one pin placed so there is only one layer that can be used. `place_pins` places multiple pins so they can be spread across different layers. `-hor_layers` and `-ver_layers` constrain the horizontal layers and vertical layers to use, which is important because these pins are being placed automatically, not manually like with `place_pin`. In theory the tool could infer the proper direction based on the layer, so I'm not sure why it has to be explicit. It might be to allow forcing of wrong way pins. @eder-matheus do you know?
3. For placing pins manually, you should simply use a long list of `place_pin` commands for each individual pin (no other commands). An easy way to do this could be to automatically place the pins and then use `write_pin_placement` to create an executable Tcl file which will individually place each pin. Alternatively if you are looking to create a very specific pin placement, you could generate a pin placement script using a generator script of some kind.

**Reply by eder-matheus** [reply]
@rovinski Actually, we do not allow wrong-way pins. PPL requires at least one vertical layer and one horizontal layer to work, so we have separate options for both layer directions. Another use for this is having a different number of layers for each direction (e.g., two layers for vertical pins and one layer for horizontal pins).

But thinking about it now, we could merge these two options into a single `-layers` option and allow using only vertical or horizontal layers. I'll keep this in mind for future PPL updates.

**Reply by b224hisl** [reply]
Thank you very much for your detailed and timely answer!!

**Reply by b224hisl** [reply]
@rovinski Yes, as you mentioned,  `define_pin_shape_pattern` is mostly used for creating pins on your top or a specific layer, it seems `place_pin` can also do such job. Here I show my pin placment after `place_pin`. 

![Pasted image 20231209225131](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/39954153/cffd0365-59e0-4f1d-b108-3a2310bb2842)

But could you please explain how `define_pin_shape_pattern` facilitate 3DIC integration of chips using face-to-face packaging technologies? Actually I'm learing this kind of topic. This command doesn't seem to change the position of any pin, nothing happens after that. I know the it just create the grid. But I'm a little confused about how to use the grid to place the pins. If I create such kind of pin grid. do I need to `place_pin` according to the grid? 

thank you very much!

**Reply by rovinski** [reply]
`define_pin_shape_pattern` is used because it creates pins directly on the top metal layer. This is not the case for more typical packaging technologies such as wirebond or flip-chip, because those technologies usually require an I/O driver cell. This requires a separate flow for placing and connecting the I/O driver cells, and in that case you should use the [`PAD` tool](https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/pad/README.md) which handles I/O cell placement, fillers, bump/pad placement, and redistribution layer (RDL) routing.

`define_pin_shape_pattern` creates the "pin placement slots" - the valid locations where pins can be placed. You need to run `place_pins` afterwards which will run the pin placement engine and actually place pins into those slots.

Again, the difference between `place_pins` and `place_pin` is that the former will automatically place all of your pins using an optimizing engine. With the latter, you have to manually place each individual pin using exact coordinates.

**Reply by b224hisl** [reply]
Sorry I get a little lost, but do you mean that `define_pin_shape_pattern` and `place_pins` should be used together? Or, after `define_pin_shape_pattern` the PAD tool should be called? 

I tried to `define_pin_shape_pattern` and then `place_pins`, but it would report the area is not enough even there is just one pin😢
Sorry, I might be a little stupid. I'm really sorry for troubling you😢

**Reply by eder-matheus** [reply]
@b224hisl  The idea is to use `define_pin_shape_pattern` together with `place_pins`. You can put individual pins in any place of the die area with the `place_pin` command, but using the `define_pin_shape_pattern` together with `place_pins` will ensure that all pins will be aligned with the routing tracks, which is essential for the detailed routing stage.

Could you share the exact error message you got from the tool? Also, if you have a reproducible of the error that you can share (not a private design or technology), it would be even more helpful to understand what's going on.

**Reply by b224hisl** [reply]
Ok, thank you very much, I think that problem has been solved, but another problem arrives. As you can see in this image, my die has its size of 250x130, then I 
 `define_pin_shape_pattern -layer met5 -x_step 2 -y_step 2 -region {64 20 185 116} -size {1 1}` and 
`define_pin_shape_pattern -layer met4 -x_step 2 -y_step 2 -region {64 20 185 116} -size {1 1}` ,
then I place pins by `place_pins -hor_layers met5 -ver_layers met4`. But the pins are still on the boundary:
![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/39954153/f1b6873b-6827-429f-aa8d-96264ae5f0dd)
 I would like to place pins like 
![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/39954153/211c1049-1c54-402a-b3b4-b0006ecf0d5d)
or
![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/39954153/67d5ef5d-64e7-49c2-a5d4-6768944c5b2c)
Are there any ways to achieve them?

**Reply by eder-matheus** [reply]
We do not support two patterns created with the `define_pin_shape_pattern` command. Only one of them will be used, probably the last one/

To have your pins placed in the grid you created, you have to constrain them to this grid with this command:
set_io_pin_constraint -pin_names * -region "up:*".

Unfortunately, our documentation on this functionality is not clear enough. I will update it to clarify how to use the grid created in the top layer.

**Reply by b224hisl** [reply]
I also execute `set_io_pin_constraint -pin_names * -region "up:*"` but it seems to get the same result, the pins are still on the boundary

**Reply by eder-matheus** [reply]
Do you have a test case you can share? Also, did you remove one of the calls for `define_pin_shape_pattern`?

**Reply by eder-matheus** [reply]
Here's one of our unit tests that performs exactly what you're trying to do: https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/ppl/test/top_layer7.tcl

**Comment by b224hisl** [comment]
OK fine, that really helps! Thank you a lot! 
Now I have a better understanding of pin placement. You can close this issue now!
But maybe I will update my latter issue related to that topic. Thanks!

**Comment by b224hisl** [comment]
However, after placing pins manually:
![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/39954153/faf18f57-104b-4184-b55b-70bd3dfa1095)
I found that there seems no routes to pins which are placed within the die, is there any way to route to these pins? 
![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/39954153/6fa41e63-aa3c-4f81-b477-fbf12e216fe3)
@eder-matheus
